server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java

```diff
@@ -612,4 +634,113 @@ private void failShard(ReplicationFailedException e, IndexShard indexShard) {
         }
     }
 
+    /**
+     * Async task responsible for failing stale replicas.
+     */
+    final static class AsyncFailStaleReplicaTask extends AbstractAsyncTask {
+        private final PendingCheckpoints pendingCheckpoints;
+        private final IndicesService indicesService;
+        static final TimeValue INTERVAL = TimeValue.timeValueSeconds(30);
+
+        public AsyncFailStaleReplicaTask(IndicesService indicesService, ThreadPool threadPool, PendingCheckpoints pendingCheckpoints) {
+            super(logger, threadPool, INTERVAL, true);
+            this.pendingCheckpoints = pendingCheckpoints;
+            this.indicesService = indicesService;
+            rescheduleIfNecessary();
+        }
+
+        @Override
+        protected boolean mustReschedule() {
+            return true;
+        }
+
+        @Override
+        protected void runInternal() {
+            List<ShardId> shardsToFail = pendingCheckpoints.getStaleShardsToFail();
+            for (ShardId shardId : shardsToFail) {
+                IndexShard replicaShard = this.indicesService.getShardOrNull(shardId);
+                if (replicaShard != null && replicaShard.state() == IndexShardState.STARTED) {
+                    failShard(new ReplicationFailedException("replica too far behind primary, marking as stale"), replicaShard);
+                }
+            }
+        }
+
+        @Override
+        protected String getThreadPool() {
+            return ThreadPool.Names.GENERIC;
+        }
+
+        @Override
+        public String toString() {
+            return "fail_stale_replica";
+        }
+
+    }
+
+    /**
+     * Internal class to track the list of Replication Checkpoints which are still not processed.
+     */
+    public static class PendingCheckpoints {
+        protected final Map<ShardId, List<Tuple<ReplicationCheckpoint, Long>>> checkpointsTracker = ConcurrentCollections
+            .newConcurrentMap();
+        private volatile TimeValue maxAllowedReplicationTime;
+
+        public PendingCheckpoints(TimeValue maxAllowedReplicationTime) {
+            this.maxAllowedReplicationTime = maxAllowedReplicationTime;
+        }
+
+        public void setMaxAllowedReplicationTime(TimeValue maxAllowedReplicationTime) {
+            this.maxAllowedReplicationTime = maxAllowedReplicationTime;
+        }
+
+        public void addNewReceivedCheckpoint(ShardId shardId, ReplicationCheckpoint checkpoint) {
+            checkpointsTracker.putIfAbsent(shardId, new ArrayList<>());
+            List<Tuple<ReplicationCheckpoint, Long>> checkpoints = checkpointsTracker.get(shardId);
+            if (checkpoints.size() > 0) {
+                assert checkpoints.get(checkpoints.size() - 1).v1().isAheadOf(checkpoint) == false;
+            }
+            checkpoints.add(new Tuple<>(checkpoint, System.currentTimeMillis()));
```

Comment:
I don't think this is thread safe. The concurrent map is thread safe but the returned ArrayList is not. To be thread safe I think you'll have to use `compute` on the Map and create new list instances on every modification (similar to what you're doing in `updateCheckpointProcessed()` below), or maybe use a thread safe List implementation, or fully `synchronize` all access to the map.

Reply:
Thanks. I've gone ahead with compute on map and returning new list on modification.